sql/pgwire: populate table and column ID in RowDescription

rafiss opened the pull request:
The RowDescription in the wire protocol contains metadata that points to
the source of the returned column. Drivers such as pgjdbc rely on this
metadata in order for certain functionality to work.

Release note (sql change): The RowDescription message of the pgwire
protocol now contains the table OID and column ID for each column in the
result set. These values correspond to pg_attribute.attrelid and
pg_attribute.attnum. If a result column does not refer to a simple table
or view, these values will be zero, as they were before.

fixes #44161

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48417)
<!-- Reviewable:end -->

jordanlewis commented on pkg/sql/pgwire/testdata/pgtest/int_size:
```diff
@@ -92,6 +93,6 @@ Query {"String": "SELECT * FROM t1"}
 until
 ReadyForQuery
 ----
```
Could you add a couple more tests in here that check what happens when a column gets threaded through a few operators?

I'm thinking we should make sure that, for example, something like `SELECT c FROM t JOIN t2 ON ... WHERE ... ORDER BY d` still tracks `c`'s table id and column id despite being munged up by all of these filter and join operators.

rafiss replied:
sure thing!

jordanlewis reviewed (approved):
LGTM overall. I just am not sure if this is "complete" yet, and adding some more tests would help find out. Nonetheless even if we didn't make it "complete" in this way it would still be worth merging.

rafiss commented:
This is ready for another review. The additional tests were helpful in finding another place to pass through this metadata.

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/copy.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gyBpw7RhjNeI1aHmb:-M6gyBpw7RhjNeI1aHmc:b-22frgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/copy.go#L137)):*
> ```Go
> 			Typ:      cols[i].Type,
> 			TableID:  tableDesc.GetID(),
> 			ColumnID: cols[i].ID}
> ```

Ooh, just realizing that these all need to be `.GetLogicalColumnID()`, since that's how `attnum` is defined now. 
___
*[pkg/sql/opt_exec_factory.go, line 1977 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6h2wkO6KbxyystNeuZ:-M6h2wkO6KbxyystNeu_:boiuhyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/opt_exec_factory.go#L1977)):*
> ```Go
> // addExpr adds a new render expression with the given name.
> func (rb *renderBuilder) addExpr(
> 	expr tree.TypedExpr, colName string, tableID sqlbase.ID, columnID sqlbase.ColumnID,
> ```

If you change the other one, columnID should also change to be called pgAttributeNum here.
___
*[pkg/sql/pgwire/conn.go, line 1287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gz1rN3nYsQousxEEu:-M6gxdZU3T9vWFmsmzKf:b9cf88q) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/conn.go#L1287)):*
> ```Go
> 		typ := pgTypeForParserType(column.Typ)
> 		c.msgBuilder.putInt32(int32(column.TableID))  // Table OID (optional).
> 		c.msgBuilder.putInt16(int16(column.ColumnID)) // Column attribute ID (optional).
> ```

This downcast is scary. I think it means that if you churn through 65k add/drops it'll break things silently...

@richardjcai @rohany do we have limits on this anywhere that you know of? Since you've been looking at this stuff.
___
*[pkg/sql/pgwire/testdata/pgtest/row_description, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6h2Zzz-vWZJP5XCH1b:-M6h2Zzz-vWZJP5XCH1c:b-2lmmy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/testdata/pgtest/row_description#L1)):*
> ```
> # This test verifies that we're populating the TableID and ColumnID in the
> ```

s/ColumnID/PGAttributeNum/
___
*[pkg/sql/pgwire/testdata/pgtest/row_description, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6h3Tb9AbM5GayyV4UM:-M6h3Tb9AbM5GayyV4UN:b10ca2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/testdata/pgtest/row_description#L80)):*
> ```
> 
> send
> Query {"String": "SELECT * FROM v WHERE v1 = 1"}
> ```

Nice tests!
___
*[pkg/sql/sqlbase/result_columns.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gy_MpASYuKGKRsspZ:-M6gy_MpASYuKGKRssp_:b-lr8t3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/sqlbase/result_columns.go#L32)):*
> ```Go
> 	// reference, these fields are zeroes.
> 	TableID  ID       // OID of column's source table (pg_attribute.attrelid).
> 	ColumnID ColumnID // Column's number in source table (pg_attribute.attnum).
> ```

I think we should change this field to be called `PGAttributeNum`, since it's actually distinct from the idea of `ColumnID` as you note.
___
*[pkg/sql/sqlbase/result_columns.go, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gyn_I6opG3Hf4BqZj:-M6gyn_I6opG3Hf4BqZk:b7m5dwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/sqlbase/result_columns.go#L58)):*
> ```Go
> 				Hidden:   hidden,
> 				TableID:  tableID,
> 				ColumnID: colDesc.ID,
> ```

ditto: This should be colDesc.GetLogicalColumnID(), how `attnum` is defined.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss, @RichardJCai, and @rohany)

___
*[pkg/sql/pgwire/conn.go, line 1287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gz1rN3nYsQousxEEu:-M6h4UGlDqa42FmbimPy:b-fqgu51) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/conn.go#L1287)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This downcast is scary. I think it means that if you churn through 65k add/drops it'll break things silently...

@richardjcai @rohany do we have limits on this anywhere that you know of? Since you've been looking at this stuff.
</blockquote></details>

it shouldn't actually be dangerous, because `pg_attribute.attnum` is an `INT2` type

but i agree, the types make it scary. `ColumnID` should be an alias for int16


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @RichardJCai, and @rohany)

___
*[pkg/sql/copy.go, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gyBpw7RhjNeI1aHmb:-M6h5FH5DWrePIBDu1Q3:bzvuo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/copy.go#L137)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Ooh, just realizing that these all need to be `.GetLogicalColumnID()`, since that's how `attnum` is defined now. 
</blockquote></details>

nice catch, i'll update them


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, and @RichardJCai)

___
*[pkg/sql/pgwire/conn.go, line 1287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6gz1rN3nYsQousxEEu:-M6h9qI6CY2mf6lr1rLI:b-bu9fcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/pgwire/conn.go#L1287)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it shouldn't actually be dangerous, because `pg_attribute.attnum` is an `INT2` type

but i agree, the types make it scary. `ColumnID` should be an alias for int16
</blockquote></details>

I don't believe that we have any tracked limits on this -- maybe we need to start keeping track


<!-- Sent from Reviewable.io -->

rafiss committed:
sql/pgwire: populate table and column ID in RowDescription

The RowDescription in the wire protocol contains metadata that points to
the source of the returned column. Drivers such as pgjdbc rely on this
metadata in order for certain functionality to work.

Release note (sql change): The RowDescription message of the pgwire
protocol now contains the table OID and column ID for each column in the
result set. These values correspond to pg_attribute.attrelid and
pg_attribute.attnum. If a result column does not refer to a simple table
or view, these values will be zero, as they were before.

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-:-M6kTxW9BwqYORDx_OgZ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/opt_exec_factory.go, line 1977 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48417#-M6h2wkO6KbxyystNeuZ:-M6kTjDR-FTxQgmcXW-W:b-stkect) ([raw file](https://github.com/cockroachdb/cockroach/blob/459ec99950ce6aa57e18db4586340ee7bdca597d/pkg/sql/opt_exec_factory.go#L1977)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

If you change the other one, columnID should also change to be called pgAttributeNum here.
</blockquote></details>

Heh, I see that GetLogicalColumnID returns sqlbase.ColumnID which feels wrong. I don't think you have to change in this PR but it's a little weird.


<!-- Sent from Reviewable.io -->

rafiss commented:
thanks for the reviews!

bors r+
